### ISSUE TITLE
Sik-Ho Tang | Review -- Multi-task Self-Supervised Visual Learning.

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#128]
Sik-Ho Tang. [Review — Multi-task Self-Supervised Visual Learning](https://sh-tsang.medium.com/review-multi-task-self-supervised-visual-learning-f7ad76bb27cb).

**Comment by NorbertZheng** [comment]
## Overview
Pretrain Using Multiple Pretext Tasks to Improve Downstream Task Accuracy.

Multi-task Self-Supervised Visual Learning. Doersch ICCV’17, by DeepMind, and VGG, University of Oxford. 2017 ICCV, Over 400 Citations.
Self-Supervised Learning, Representation Learning, Image Classification, Object Detection, Depth Prediction.

 - In this paper, 4 different self-supervised tasks using [ResNet](https://towardsdatascience.com/resnet-with-identity-mapping-over-1000-layers-reached-image-classification-bb50a42af03e)-101 are combined to jointly train a network.
 - **Lasso regularization is used to factorize the information in its representation**, and **methods for “harmonizing” network inputs** in order to learn a more unified representation.

**Comment by NorbertZheng** [comment]
A joint loss that includes 4 task losses???

**Comment by NorbertZheng** [comment]
## Multi-Task Network
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/8334c833-f922-49e7-bbf3-944107fe63a5)
The structure of our multi-task network based on [ResNet](https://towardsdatascience.com/resnet-with-identity-mapping-over-1000-layers-reached-image-classification-bb50a42af03e)-101, with block 3 having 23 residual units. a) Naïve shared-trunk approach, where each “head” is attached to the output of block 3. b) the lasso architecture, where each “head” receives a linear combination of unit outputs within block3, weighted by the matrix , which is trained to be sparse.

Three architectures are described:
 - First, the (naïve) multi-task network that has a common trunk and a head for each task (figure a). I.e. just split the embedding of the last layer to get multiple heads.
 - Second, the lasso extension of this architecture (figure b) that enables the training to **determine the combination of layers to use for each self-supervised task**.
 - Third, a method for harmonizing input channels across self-supervision tasks.

**Comment by NorbertZheng** [comment]
### Common Trunk
Model:
 - [ResNet](https://towardsdatascience.com/resnet-with-identity-mapping-over-1000-layers-reached-image-classification-bb50a42af03e)-101, as in [Pre-Activation ResNet](https://towardsdatascience.com/resnet-with-identity-mapping-over-1000-layers-reached-image-classification-bb50a42af03e), is used. The entire architecture up to the end of block 3 are kept.
 - This same block3 representation is used to solve all tasks and evaluations (see figure a). Thus, the “trunk” has an output with 1024 channels, and consists of 88 convolution layers with roughly 30 million parameters.

One embedding for all tasks!!!

Each task has a separate loss, and has extra layers in a “head,” which may have a complicated structure.

Implementation:
 - For instance, the relative position ([Context Prediction](https://sh-tsang.medium.com/review-unsupervised-visual-representation-learning-by-context-prediction-self-supervised-51a1d7ce6aff)) and [Exemplar](https://sh-tsang.medium.com/review-exemplar-cnn-discriminative-unsupervised-feature-learning-with-convolutional-neural-fa68abe937cc) tasks have a Siamese architecture. This is achieved by passing all patches through the trunk as a single batch, and then rearranging the elements in the batch to make pairs (or triplets) of representations to be processed by the head.
 - At each training iteration, only one of the heads is active. However, gradients are averaged across many iterations where different heads are active, meaning that **the overall loss is a sum of the losses of different tasks**.

4 Self-supervised tasks are used: Relative Position ([Context Prediction](https://sh-tsang.medium.com/review-unsupervised-visual-representation-learning-by-context-prediction-self-supervised-51a1d7ce6aff)), [Colorization](https://sh-tsang.medium.com/review-colorful-image-colorization-self-supervised-learning-c2667640bd0a), [Exemplar](https://sh-tsang.medium.com/review-exemplar-cnn-discriminative-unsupervised-feature-learning-with-convolutional-neural-fa68abe937cc), and Motion Segmentation ([Motion Masks](https://sh-tsang.medium.com/review-motion-masks-learning-features-by-watching-objects-move-548e7d955c2e)).

**Comment by NorbertZheng** [comment]
One task one batch, average across all tasks to get an averaged gradient with lower variance!!!

**Comment by NorbertZheng** [comment]
### Separating Features via Lasso
Idea:
 - **Different tasks require different features.**

If the features are factorized into different tasks, then the network can select from the discovered feature groups while training on the evaluation tasks.

Model:
 - A linear combination of skip layers is passed to each head. Concretely, each task has a set of coefficients, one for each of the 23 candidate layers in block 3. The representation that’s fed into each task head is a sum of the layer activations weighted by these task-specific coefficients.
 - **A lasso (L1) penalty is proposed to encourage the combination to be sparse, which therefore encourages the network to concentrate all of the information required by a single task into a small number of layers.**

One task involves as fewer embedding layers as possible.

 - **Thus, when fine-tuning on a new task, these task-specific layers can be quickly selected or rejected as a group.**

The representation passed to the head for task n is then:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/093bc557-1824-46d2-8083-4fd554695060)
where $N$ is the number of self-supervised tasks, $M$ is the number of residual units in block 3, and $Unit_{m}$ is the output of residual unit $m$.

To ensure sparsity, an L1 penalty is added on the entries of $\alpha$ to the objective function. A similar $\alpha$ matrix is created for the set of evaluation tasks.

**Comment by NorbertZheng** [comment]
One task is a combination of different embedding layers (across different abstraction levels!!!)

**Comment by NorbertZheng** [comment]
### Harmonizing Network Inputs
 - **Each self-supervised task pre-processes its data differently.**

To “harmonize,” relative position ([Context Prediction](https://sh-tsang.medium.com/review-unsupervised-visual-representation-learning-by-context-prediction-self-supervised-51a1d7ce6aff))’s preprocessing is replaced with the same preprocessing used for [Colorization](https://sh-tsang.medium.com/review-colorful-image-colorization-self-supervised-learning-c2667640bd0a): images are converted to Lab, and the a and b channels are discarded (The L channel is replicated by 3 times).

**Comment by NorbertZheng** [comment]
### Distributed Network Training
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/f0b01733-a59d-4c7d-8460-32c53f82bc66)
Distributed training setup.

Training Setup:
 - Each machine trains the network on a single task.
 - Several GPU machines are allocated for each task, and gradients from each task are synchronized and aggregated with separate RMSProp optimizers.
 - A hybrid approach is used that **gradients are accumulated from all workers that are working on a single task**, and then have the parameter servers apply the aggregated gradients from a single task when ready, **without synchronizing with other tasks**.

Results:
 - Experiments found that this approach resulted in faster learning than either purely synchronous or purely asynchronous training, and in particular, was more stable than asynchronous training.

64 GPUs are used in parallel, and checkpoints are saved every roughly 2.4K GPU (NVIDIA K40) hours.

**Comment by NorbertZheng** [comment]
Synchronizing within the same task, while ont synchronizing with other tasks.

**Comment by NorbertZheng** [comment]
## Model Fine-Tuning
### ImageNet
After self-supervised training, a single linear classification layer (a softmax) to the network at the end of block 3, and train on the full ImageNet training set.

 - **All pre-trained weights are frozen during training.**

### PASCAL VOC 2007 Detection
[Fast R-CNN](https://medium.com/coinmonks/review-fast-r-cnn-object-detection-a82e172e87ba) is used, which trains a single network base with multiple heads (common trunk, more stable???) for object proposals, box classification, and box localization.

 - **All network weights are fine-tuned.**

### NYU V2 Depth Prediction
[ResNet](https://towardsdatascience.com/resnet-with-identity-mapping-over-1000-layers-reached-image-classification-bb50a42af03e)-50 is used. The block 3 outputs are directly fed into the up-projection layers. I.e. append other decoders after the pre-trained model.

 - **All network weights are fine-tuned.**

**Comment by NorbertZheng** [comment]
## Experimental Results
### Individual Self-Supervised Training Performance
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/ba95acb7-7f65-407b-bfeb-d225789ff65a)
Individual Self-Supervised Training Performance.

Results:
 - Of the self-supervised pre-training methods, relative position ([Context Prediction](https://sh-tsang.medium.com/review-unsupervised-visual-representation-learning-by-context-prediction-self-supervised-51a1d7ce6aff)) and [Colorization](https://sh-tsang.medium.com/review-colorful-image-colorization-self-supervised-learning-c2667640bd0a) are the top performers, with relative position ([Context Prediction](https://sh-tsang.medium.com/review-unsupervised-visual-representation-learning-by-context-prediction-self-supervised-51a1d7ce6aff)) winning on PASCAL and NYU, and [Colorization](https://sh-tsang.medium.com/review-colorful-image-colorization-self-supervised-learning-c2667640bd0a) winning on ImageNet-frozen.
 - Remarkably, relative position ([Context Prediction](https://sh-tsang.medium.com/review-unsupervised-visual-representation-learning-by-context-prediction-self-supervised-51a1d7ce6aff)) performs on-par with ImageNet pre-training on depth prediction, and the gap is just 7.5% mAP on PASCAL.
 - The only task where the gap remains large is the ImageNet evaluation itself, which is not surprising since the ImageNet pretraining and evaluation use the same labels.
 - Motion segmentation ([Motion Masks](https://sh-tsang.medium.com/review-motion-masks-learning-features-by-watching-objects-move-548e7d955c2e)) and [Exemplar](https://sh-tsang.medium.com/review-exemplar-cnn-discriminative-unsupervised-feature-learning-with-convolutional-neural-fa68abe937cc) training are somewhat worse than the others, with [Exemplar](https://sh-tsang.medium.com/review-exemplar-cnn-discriminative-unsupervised-feature-learning-with-convolutional-neural-fa68abe937cc) worst on Pascal and NYU, and motion segmentation ([Motion Masks](https://sh-tsang.medium.com/review-motion-masks-learning-features-by-watching-objects-move-548e7d955c2e)) worst on ImageNet.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/9cd14321-05fd-4798-b93c-342c0e6a39d6)
Comparison of performance for different self-supervised methods over time.

 - After 16.8K GPU hours, performance is plateauing but has not completely saturated.
 - Interestingly, on the ImageNet frozen evaluation, where [Colorization](https://sh-tsang.medium.com/review-colorful-image-colorization-self-supervised-learning-c2667640bd0a) is winning, the gap relative to relative position ([Context Prediction](https://sh-tsang.medium.com/review-unsupervised-visual-representation-learning-by-context-prediction-self-supervised-51a1d7ce6aff)) is growing.
 - Also, while most algorithms slowly improve performance with training time, [Exemplar](https://sh-tsang.medium.com/review-exemplar-cnn-discriminative-unsupervised-feature-learning-with-convolutional-neural-fa68abe937cc) training doesn’t fit this pattern.

**Comment by NorbertZheng** [comment]
- Pretraining on ImageNet with supervised learning signal, and evaluating performance on the same dataset (or domain) always performs the best!!!
 - Supervised learning signal is better than un-supervised learning signal when dataset is relatively small (compared to GPT-4) when executing cross-domain tasks.
 - The overall pretraining method seems like an ensemble of different pretraining methods!!!

**Comment by NorbertZheng** [comment]
### Naïve Multi-Task Combination of Self-Supervision Tasks
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/0033fce7-99df-4d3f-924b-35ae71261ad9)
Comparison of various combinations of self-supervised tasks RP: Relative Position ([Context Prediction](https://sh-tsang.medium.com/review-unsupervised-visual-representation-learning-by-context-prediction-self-supervised-51a1d7ce6aff)); Col: [Colorization](https://sh-tsang.medium.com/review-colorful-image-colorization-self-supervised-learning-c2667640bd0a); Ex: [Exemplar](https://sh-tsang.medium.com/review-exemplar-cnn-discriminative-unsupervised-feature-learning-with-convolutional-neural-fa68abe937cc) Nets; MS: Motion Segmentation ([Motion Masks](https://sh-tsang.medium.com/review-motion-masks-learning-features-by-watching-objects-move-548e7d955c2e)). Metrics: ImageNet: Recall@5; PASCAL: mAP; NYU: % Pixels below 1.25.

 - Adding either [Colorization](https://sh-tsang.medium.com/review-colorful-image-colorization-self-supervised-learning-c2667640bd0a) or [Exemplar](https://sh-tsang.medium.com/review-exemplar-cnn-discriminative-unsupervised-feature-learning-with-convolutional-neural-fa68abe937cc) leads to more than 6 points gain on ImageNet.
 - Adding both [Colorization](https://sh-tsang.medium.com/review-colorful-image-colorization-self-supervised-learning-c2667640bd0a) and [Exemplar](https://sh-tsang.medium.com/review-exemplar-cnn-discriminative-unsupervised-feature-learning-with-convolutional-neural-fa68abe937cc) gives a further 2% boost.
 - The best-performing method was a combination of all four self-supervised tasks.

**Comment by NorbertZheng** [comment]
More tasks!!!

**Comment by NorbertZheng** [comment]
### Harmonization
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/a8211f12-2fd6-486c-8280-6e6e1af4ccd1)
Comparison of methods with and without harmonization, H: harmonization.

 - There are large improvements on ImageNet.
 - The other two evaluation tasks do not show any improvement with harmonization.
 - **This suggests that our networks are actually quite good at dealing with stark differences between pre-training data domains when the features are fine-tuned at test time.**

**Comment by NorbertZheng** [comment]
Data formation matters (RP compared to RP/H on ImageNet), but only when the amount of the dataset is relatively small. When the amount of dataset is large, the gap will decrease (RP compared to RP/H on PASCAL/NYU)!!!

**Comment by NorbertZheng** [comment]
### Lasso
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/996b2256-17a6-48be-905b-d1d49ebbe2bc)
Comparison of performance with and without the lasso technique for factorizing representations, for a network trained on all four self-supervised tasks for 16.8K GPU-hours.

There are four cases: no lasso, lasso only on the evaluation tasks, lasso only at pre-training time, and lasso in both self-supervised training and evaluation.

**Comment by NorbertZheng** [comment]
The gap between ImageNet pre-trained and self-supervision pre-trained with four tasks is nearly closed for the VOC detection evaluation, and completely closed for NYU depth.

**Comment by NorbertZheng** [comment]
## Reference
 - [2017 ICCV] [Doersch ICCV’17] [Multi-task Self-Supervised Visual Learning](https://arxiv.org/abs/1708.07860).
